dbms/src/DataStreams/HashJoinProbeBlockInputStream.cpp

```diff
@@ -234,43 +114,43 @@ Block HashJoinProbeBlockInputStream::getOutputBlock()
     {
         while (true)
         {
+            if unlikely (isCancelledOrThrowIfKilled())
+                return {};
+
             switch (status)
             {
             case ProbeStatus::WAIT_BUILD_FINISH:
-                join->waitUntilAllBuildFinished();
+            {
+                auto & cur_probe_exec = *probe_exec;
+                cur_probe_exec.waitUntilAllBuildFinished();
                 /// after Build finish, always go to Probe stage
-                if (join->isRestoreJoin())
-                    current_probe_stream->readSuffix();
-                status = ProbeStatus::PROBE;
+                cur_probe_exec.onProbeStart();
+                switchStatus(ProbeStatus::PROBE);
                 break;
+            }
             case ProbeStatus::PROBE:
             {
-                assert(current_probe_stream != nullptr);
-                if (probe_process_info.all_rows_joined_finish)
+                auto ret = probe_exec->probe();
```

Comment:
Why sometimes `probe_exec` is used directly, and sometimes `probe_exec` is used as `auto & cur_probe_exec = *probe_exec;`?

Reply:
Just to avoid duplicate fetches from probe_exec because of the locking overhead